File: services/google/appengine/application.go
// Copyright 2021 Google LLC. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package appengine

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"

	"google.golang.org/api/googleapi"
	"github.com/GoogleCloudPlatform/declarative-resource-client-library/dcl"
)

type Application struct {
	AuthDomain      *string                       `json:"authDomain"`
	CodeBucket      *string                       `json:"codeBucket"`
	DatabaseType    *ApplicationDatabaseTypeEnum  `json:"databaseType"`
	DefaultBucket   *string                       `json:"defaultBucket"`
	DefaultHostname *string                       `json:"defaultHostname"`
	DispatchRules   []ApplicationDispatchRules    `json:"dispatchRules"`
	FeatureSettings *ApplicationFeatureSettings   `json:"featureSettings"`
	GcrDomain       *string                       `json:"gcrDomain"`
	Iap             *ApplicationIap               `json:"iap"`
	Name            *string                       `json:"name"`
	Location        *string                       `json:"location"`
	ServingStatus   *ApplicationServingStatusEnum `json:"servingStatus"`
}

func (r *Application) String() string {
	return dcl.SprintResource(r)
}

// The enum ApplicationDatabaseTypeEnum.
type ApplicationDatabaseTypeEnum string

// ApplicationDatabaseTypeEnumRef returns a *ApplicationDatabaseTypeEnum with the value of string s
// If the empty string is provided, nil is returned.
func ApplicationDatabaseTypeEnumRef(s string) *ApplicationDatabaseTypeEnum {
	if s == "" {
		return nil
	}

	v := ApplicationDatabaseTypeEnum(s)
	return &v
}

func (v ApplicationDatabaseTypeEnum) Validate() error {
	for _, s := range []string{"DATABASE_TYPE_UNSPECIFIED", "CLOUD_DATASTORE", "CLOUD_FIRESTORE", "CLOUD_DATASTORE_COMPATIBILITY"} {
		if string(v) == s {
			return nil
		}
	}
	return &dcl.EnumInvalidError{
		Enum:  "ApplicationDatabaseTypeEnum",
		Value: string(v),
		Valid: []string{},
	}
}

// The enum ApplicationServingStatusEnum.
type ApplicationServingStatusEnum string

// ApplicationServingStatusEnumRef returns a *ApplicationServingStatusEnum with the value of string s
// If the empty string is provided, nil is returned.
func ApplicationServingStatusEnumRef(s string) *ApplicationServingStatusEnum {
	if s == "" {
		return nil
	}

	v := ApplicationServingStatusEnum(s)
	return &v
}

func (v ApplicationServingStatusEnum) Validate() error {
	for _, s := range []string{"UNSPECIFIED", "SERVING", "USER_DISABLED", "SYSTEM_DISABLED"} {
		if string(v) == s {
			return nil
		}
	}
	return &dcl.EnumInvalidError{
		Enum:  "ApplicationServingStatusEnum",
		Value: string(v),
		Valid: []string{},
	}
}

type ApplicationDispatchRules struct {
	empty   bool    `json:"-"`
	Domain  *string `json:"domain"`
	Path    *string `json:"path"`
	Service *string `json:"service"`
}

type jsonApplicationDispatchRules ApplicationDispatchRules

func (r *ApplicationDispatchRules) UnmarshalJSON(data []byte) error {
	var res jsonApplicationDispatchRules
	if err := json.Unmarshal(data, &res); err != nil {
		return err
	}

	var m map[string]interface{}
	json.Unmarshal(data, &m)

	if len(m) == 0 {
		*r = *EmptyApplicationDispatchRules
	} else {

		r.Domain = res.Domain

		r.Path = res.Path

		r.Service = res.Service

	}
	return nil
}

// This object is used to assert a desired state where this ApplicationDispatchRules is
// empty.  Go lacks global const objects, but this object should be treated
// as one.  Modifying this object will have undesirable results.
var EmptyApplicationDispatchRules *ApplicationDispatchRules = &ApplicationDispatchRules{empty: true}

func (r *ApplicationDispatchRules) String() string {
	return dcl.SprintResource(r)
}

func (r *ApplicationDispatchRules) HashCode() string {
	// Placeholder for a more complex hash method that handles ordering, etc
	// Hash resource body for easy comparison later
	hash := sha256.New().Sum([]byte(r.String()))
	return fmt.Sprintf("%x", hash)
}

type ApplicationFeatureSettings struct {
	empty                   bool  `json:"-"`
	SplitHealthChecks       *bool `json:"splitHealthChecks"`
	UseContainerOptimizedOs *bool `json:"useContainerOptimizedOs"`
}

type jsonApplicationFeatureSettings ApplicationFeatureSettings

func (r *ApplicationFeatureSettings) UnmarshalJSON(data []byte) error {
	var res jsonApplicationFeatureSettings
	if err := json.Unmarshal(data, &res); err != nil {
		return err
	}

	var m map[string]interface{}
	json.Unmarshal(data, &m)

	if len(m) == 0 {
		*r = *EmptyApplicationFeatureSettings
	} else {

		r.SplitHealthChecks = res.SplitHealthChecks

		r.UseContainerOptimizedOs = res.UseContainerOptimizedOs

	}
	return nil
}

// This object is used to assert a desired state where this ApplicationFeatureSettings is
// empty.  Go lacks global const objects, but this object should be treated
// as one.  Modifying this object will have undesirable results.
var EmptyApplicationFeatureSettings *ApplicationFeatureSettings = &ApplicationFeatureSettings{empty: true}

func (r *ApplicationFeatureSettings) String() string {
	return dcl.SprintResource(r)
}

func (r *ApplicationFeatureSettings) HashCode() string {
	// Placeholder for a more complex hash method that handles ordering, etc
	// Hash resource body for easy comparison later
	hash := sha256.New().Sum([]byte(r.String()))
	return fmt.Sprintf("%x", hash)
}

type ApplicationIap struct {
	empty                    bool    `json:"-"`
	Enabled                  *bool   `json:"enabled"`
	OAuth2ClientId           *string `json:"oauth2ClientId"`
	OAuth2ClientSecret       *string `json:"oauth2ClientSecret"`
	OAuth2ClientSecretSha256 *string `json:"oauth2ClientSecretSha256"`
}

type jsonApplicationIap ApplicationIap

func (r *ApplicationIap) UnmarshalJSON(data []byte) error {
	var res jsonApplicationIap
	if err := json.Unmarshal(data, &res); err != nil {
		return err
	}

	var m map[string]interface{}
	json.Unmarshal(data, &m)

	if len(m) == 0 {
		*r = *EmptyApplicationIap
	} else {

		r.Enabled = res.Enabled

		r.OAuth2ClientId = res.OAuth2ClientId

		r.OAuth2ClientSecret = res.OAuth2ClientSecret

		r.OAuth2ClientSecretSha256 = res.OAuth2ClientSecretSha256

	}
	return nil
}

// This object is used to assert a desired state where this ApplicationIap is
// empty.  Go lacks global const objects, but this object should be treated
// as one.  Modifying this object will have undesirable results.
var EmptyApplicationIap *ApplicationIap = &ApplicationIap{empty: true}

func (r *ApplicationIap) String() string {
	return dcl.SprintResource(r)
}

func (r *ApplicationIap) HashCode() string {
	// Placeholder for a more complex hash method that handles ordering, etc
	// Hash resource body for easy comparison later
	hash := sha256.New().Sum([]byte(r.String()))
	return fmt.Sprintf("%x", hash)
}

// Describe returns a simple description of this resource to ensure that automated tools
// can identify it.
func (r *Application) Describe() dcl.ServiceTypeVersion {
	return dcl.ServiceTypeVersion{
		Service: "app_engine",
		Type:    "Application",
		Version: "appengine",
	}
}

const ApplicationMaxPage = -1

type ApplicationList struct {
	Items []*Application

	nextToken string
}

func (c *Client) GetApplication(ctx context.Context, r *Application) (*Application, error) {
	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	b, err := c.getApplicationRaw(ctx, r)
	if err != nil {
		if dcl.IsNotFoundOrCode(err, 403) {
			return nil, &googleapi.Error{
				Code:    404,
				Message: err.Error(),
			}
		}
		return nil, err
	}
	result, err := unmarshalApplication(b, c)
	if err != nil {
		return nil, err
	}
	result.Name = r.Name

	c.Config.Logger.Infof("Retrieved raw result state: %v", result)
	c.Config.Logger.Infof("Canonicalizing with specified state: %v", r)
	result, err = canonicalizeApplicationNewState(c, result, r)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Created result state: %v", result)

	return result, nil
}

func (c *Client) ApplyApplication(ctx context.Context, rawDesired *Application, opts ...dcl.ApplyOption) (*Application, error) {
	c.Config.Logger.Info("Beginning ApplyApplication...")
	c.Config.Logger.Infof("User specified desired state: %v", rawDesired)

	ctx, cancel := context.WithTimeout(ctx, c.Config.Timeout)
	defer cancel()

	// 1.1: Validation of user-specified fields in desired state.
	if err := rawDesired.validate(); err != nil {
		return nil, err
	}

	initial, desired, diffs, err := c.applicationDiffsForRawDesired(ctx, rawDesired, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to create a diff: %w", err)
	}

	// TODO(magic-modules-eng): 2.2 Feasibility check (all updates are feasible so far).

	// 2.3: Lifecycle Directive Check
	var create bool
	var recreate bool
	lp := dcl.FetchLifecycleParams(opts)
	if initial == nil {
		if dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
			return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Creation blocked by lifecycle params: %#v.", desired)}
		}
		create = true
	} else if dcl.HasLifecycleParam(lp, dcl.BlockAcquire) {
		return nil, dcl.ApplyInfeasibleError{
			Message: fmt.Sprintf("Resource already exists - apply blocked by lifecycle params: %#v.", initial),
		}
	} else {
		for _, d := range diffs {
			if d.RequiresRecreate {
				if dcl.HasLifecycleParam(lp, dcl.BlockDestruction) || dcl.HasLifecycleParam(lp, dcl.BlockCreation) {
					return nil, dcl.ApplyInfeasibleError{
						Message: fmt.Sprintf("Infeasible update: (%v) would require recreation.", d),
					}
				}
				c.Config.Logger.Infof("Diff requires recreate: %+v\n", d)
				recreate = true
			}
			if dcl.HasLifecycleParam(lp, dcl.BlockModification) {
				return nil, dcl.ApplyInfeasibleError{Message: fmt.Sprintf("Modification blocked, diff (%v) unresolvable.", d)}
			}
		}
	}

	// 2.4 Imperative Request Planning
	var ops []applicationApiOperation
	if create {
		ops = append(ops, &createApplicationOperation{})
	} else if recreate {

		ops = append(ops, &createApplicationOperation{})
		// We should re-canonicalize based on a nil existing resource.
		desired, err = canonicalizeApplicationDesiredState(rawDesired, nil)
		if err != nil {
			return nil, err
		}
	} else {
		for _, d := range diffs {
			ops = append(ops, d.UpdateOp)
		}
	}
	c.Config.Logger.Infof("Created plan: %#v", ops)

	// 2.5 Request Actuation
	for _, op := range ops {
		c.Config.Logger.Infof("Performing operation %#v", op)
		if err := op.do(ctx, desired, c); err != nil {
			c.Config.Logger.Infof("Failed operation %#v: %v", op, err)
			return nil, err
		}
		c.Config.Logger.Infof("Finished operation %#v", op)
	}

	// 3.1, 3.2a Retrieval of raw new state & canonicalization with desired state
	c.Config.Logger.Info("Retrieving raw new state...")
	rawNew, err := c.GetApplication(ctx, desired.urlNormalized())
	if err != nil {
		return nil, err
	}

	// Get additional values from the first response.
	// These values should be merged into the newState above.
	if len(ops) > 0 {
		lastOp := ops[len(ops)-1]
		if o, ok := lastOp.(*createApplicationOperation); ok {
			if r, hasR := o.FirstResponse(); hasR {

				c.Config.Logger.Info("Retrieving raw new state from operation...")

				fullResp, err := unmarshalMapApplication(r, c)
				if err != nil {
					return nil, err
				}

				rawNew, err = canonicalizeApplicationNewState(c, rawNew, fullResp)
				if err != nil {
					return nil, err
				}
			}
		}
	}

	c.Config.Logger.Infof("Canonicalizing with raw desired state: %v", rawDesired)
	// 3.2b Canonicalization of raw new state using raw desired state
	newState, err := canonicalizeApplicationNewState(c, rawNew, rawDesired)
	if err != nil {
		return nil, err
	}

	c.Config.Logger.Infof("Created canonical new state: %v", newState)
	// 3.3 Comparison of the new state and raw desired state.
	// TODO(magic-modules-eng): EVENTUALLY_CONSISTENT_UPDATE
	newDesired, err := canonicalizeApplicationDesiredState(rawDesired, newState)
	if err != nil {
		return nil, err
	}
	c.Config.Logger.Infof("Diffing using canonicalized desired state: %v", newDesired)
	newDiffs, err := diffApplication(c, newDesired, newState)
	if err != nil {
		return nil, err
	}

	if len(newDiffs) == 0 {
		c.Config.Logger.Info("No diffs found. Apply was successful.")
	} else {
		c.Config.Logger.Infof("Found diffs: %v", newDiffs)
		diffMessages := make([]string, len(newDiffs))
		for i, d := range newDiffs {
			diffMessages[i] = fmt.Sprintf("%v", d)
		}
		return newState, dcl.DiffAfterApplyError{Diffs: diffMessages}
	}
	c.Config.Logger.Info("Done Apply.")
	return newState, nil
}

File: python/services/appengine/application.py
# Copyright 2021 Google LLC. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from connector import channel
from google3.cloud.graphite.mmv2.services.google.appengine import application_pb2
from google3.cloud.graphite.mmv2.services.google.appengine import application_pb2_grpc

from typing import List


class Application(object):
    def __init__(
        self,
        auth_domain: str = None,
        blocked_addresses: list = None,
        code_bucket: str = None,
        consumer_project: dict = None,
        database_type: str = None,
        default_bucket: str = None,
        default_hostname: str = None,
        dispatch_rules: list = None,
        domains: list = None,
        feature_settings: dict = None,
        gcr_domain: str = None,
        iap: dict = None,
        parent: dict = None,
        name: str = None,
        location: str = None,
        serving_status: str = None,
        service_account: str = None,
        service_account_file: str = "",
    ):

        channel.initialize()
        self.auth_domain = auth_domain
        self.blocked_addresses = blocked_addresses
        self.code_bucket = code_bucket
        self.consumer_project = consumer_project
        self.database_type = database_type
        self.default_bucket = default_bucket
        self.default_hostname = default_hostname
        self.dispatch_rules = dispatch_rules
        self.domains = domains
        self.feature_settings = feature_settings
        self.gcr_domain = gcr_domain
        self.iap = iap
        self.parent = parent
        self.name = name
        self.location = location
        self.serving_status = serving_status
        self.service_account = service_account
        self.service_account_file = service_account_file

    def apply(self):
        stub = application_pb2_grpc.AppengineApplicationServiceStub(channel.Channel())
        request = application_pb2.ApplyAppengineApplicationRequest()
        if Primitive.to_proto(self.auth_domain):
            request.resource.auth_domain = Primitive.to_proto(self.auth_domain)

        if Primitive.to_proto(self.blocked_addresses):
            request.resource.blocked_addresses.extend(
                Primitive.to_proto(self.blocked_addresses)
            )
        if Primitive.to_proto(self.code_bucket):
            request.resource.code_bucket = Primitive.to_proto(self.code_bucket)

        if ApplicationConsumerProject.to_proto(self.consumer_project):
            request.resource.consumer_project.CopyFrom(
                ApplicationConsumerProject.to_proto(self.consumer_project)
            )
        else:
            request.resource.ClearField("consumer_project")
        if ApplicationDatabaseTypeEnum.to_proto(self.database_type):
            request.resource.database_type = ApplicationDatabaseTypeEnum.to_proto(
                self.database_type
            )

        if Primitive.to_proto(self.default_bucket):
            request.resource.default_bucket = Primitive.to_proto(self.default_bucket)

        if Primitive.to_proto(self.default_hostname):
            request.resource.default_hostname = Primitive.to_proto(
                self.default_hostname
            )

        if ApplicationDispatchRulesArray.to_proto(self.dispatch_rules):
            request.resource.dispatch_rules.extend(
                ApplicationDispatchRulesArray.to_proto(self.dispatch_rules)
            )
        if Primitive.to_proto(self.domains):
            request.resource.domains.extend(Primitive.to_proto(self.domains))
        if ApplicationFeatureSettings.to_proto(self.feature_settings):
            request.resource.feature_settings.CopyFrom(
                ApplicationFeatureSettings.to_proto(self.feature_settings)
            )
        else:
            request.resource.ClearField("feature_settings")
        if Primitive.to_proto(self.gcr_domain):
            request.resource.gcr_domain = Primitive.to_proto(self.gcr_domain)

        if ApplicationIap.to_proto(self.iap):
            request.resource.iap.CopyFrom(ApplicationIap.to_proto(self.iap))
        else:
            request.resource.ClearField("iap")
        if ApplicationParent.to_proto(self.parent):
            request.resource.parent.CopyFrom(ApplicationParent.to_proto(self.parent))
        else:
            request.resource.ClearField("parent")
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        if ApplicationServingStatusEnum.to_proto(self.serving_status):
            request.resource.serving_status = ApplicationServingStatusEnum.to_proto(
                self.serving_status
            )

        if Primitive.to_proto(self.service_account):
            request.resource.service_account = Primitive.to_proto(self.service_account)

        request.service_account_file = self.service_account_file

        response = stub.ApplyAppengineApplication(request)
        self.auth_domain = Primitive.from_proto(response.auth_domain)
        self.blocked_addresses = Primitive.from_proto(response.blocked_addresses)
        self.code_bucket = Primitive.from_proto(response.code_bucket)
        self.consumer_project = ApplicationConsumerProject.from_proto(
            response.consumer_project
        )
        self.database_type = ApplicationDatabaseTypeEnum.from_proto(
            response.database_type
        )
        self.default_bucket = Primitive.from_proto(response.default_bucket)
        self.default_hostname = Primitive.from_proto(response.default_hostname)
        self.dispatch_rules = ApplicationDispatchRulesArray.from_proto(
            response.dispatch_rules
        )
        self.domains = Primitive.from_proto(response.domains)
        self.feature_settings = ApplicationFeatureSettings.from_proto(
            response.feature_settings
        )
        self.gcr_domain = Primitive.from_proto(response.gcr_domain)
        self.iap = ApplicationIap.from_proto(response.iap)
        self.parent = ApplicationParent.from_proto(response.parent)
        self.name = Primitive.from_proto(response.name)
        self.location = Primitive.from_proto(response.location)
        self.serving_status = ApplicationServingStatusEnum.from_proto(
            response.serving_status
        )
        self.service_account = Primitive.from_proto(response.service_account)

    @classmethod
    def delete(self, service_account_file=""):
        stub = application_pb2_grpc.AppengineApplicationServiceStub(channel.Channel())
        request = application_pb2.DeleteAppengineApplicationRequest()
        request.service_account_file = service_account_file
        response = stub.DeleteAppengineApplication(request)

    @classmethod
    def list(self, service_account_file=""):
        stub = application_pb2_grpc.AppengineApplicationServiceStub(channel.Channel())
        request = application_pb2.ListAppengineApplicationRequest()
        request.service_account_file = service_account_file

        return stub.ListAppengineApplication(request).items

    @classmethod
    def from_any(self, any_proto):
        # Marshal any proto to regular proto.
        res_proto = application_pb2.AppengineApplication()
        any_proto.Unpack(res_proto)

        res = Application()
        res.auth_domain = Primitive.from_proto(res_proto.auth_domain)
        res.blocked_addresses = Primitive.from_proto(res_proto.blocked_addresses)
        res.code_bucket = Primitive.from_proto(res_proto.code_bucket)
        res.consumer_project = ApplicationConsumerProject.from_proto(
            res_proto.consumer_project
        )
        res.database_type = ApplicationDatabaseTypeEnum.from_proto(
            res_proto.database_type
        )
        res.default_bucket = Primitive.from_proto(res_proto.default_bucket)
        res.default_hostname = Primitive.from_proto(res_proto.default_hostname)
        res.dispatch_rules = ApplicationDispatchRulesArray.from_proto(
            res_proto.dispatch_rules
        )
        res.domains = Primitive.from_proto(res_proto.domains)
        res.feature_settings = ApplicationFeatureSettings.from_proto(
            res_proto.feature_settings
        )
        res.gcr_domain = Primitive.from_proto(res_proto.gcr_domain)
        res.iap = ApplicationIap.from_proto(res_proto.iap)
        res.parent = ApplicationParent.from_proto(res_proto.parent)
        res.name = Primitive.from_proto(res_proto.name)
        res.location = Primitive.from_proto(res_proto.location)
        res.serving_status = ApplicationServingStatusEnum.from_proto(
            res_proto.serving_status
        )
        res.service_account = Primitive.from_proto(res_proto.service_account)
        return res


class ApplicationConsumerProject(object):
    def __init__(
        self,
        service: str = None,
        name: str = None,
        project_id: str = None,
        route_hash: dict = None,
        region: str = None,
    ):
        self.service = service
        self.name = name
        self.project_id = project_id
        self.route_hash = route_hash
        self.region = region

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = application_pb2.AppengineApplicationConsumerProject()
        if ApplicationConsumerProjectServiceEnum.to_proto(resource.service):
            res.service = ApplicationConsumerProjectServiceEnum.to_proto(
                resource.service
            )
        if Primitive.to_proto(resource.name):
            res.name = Primitive.to_proto(resource.name)
        if Primitive.to_proto(resource.project_id):
            res.project_id = Primitive.to_proto(resource.project_id)
        if ApplicationConsumerProjectRouteHash.to_proto(resource.route_hash):
            res.route_hash.CopyFrom(
                ApplicationConsumerProjectRouteHash.to_proto(resource.route_hash)
            )
        else:
            res.ClearField("route_hash")
        if Primitive.to_proto(resource.region):
            res.region = Primitive.to_proto(resource.region)
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ApplicationConsumerProject(
            service=resource.service,
            name=resource.name,
            project_id=resource.project_id,
            route_hash=resource.route_hash,
            region=resource.region,
        )


class ApplicationConsumerProjectArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ApplicationConsumerProject.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ApplicationConsumerProject.from_proto(i) for i in resources]


class ApplicationConsumerProjectRouteHash(object):
    def __init__(self, project_hash: str = None):
        self.project_hash = project_hash

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = application_pb2.AppengineApplicationConsumerProjectRouteHash()
        if Primitive.to_proto(resource.project_hash):
            res.project_hash = Primitive.to_proto(resource.project_hash)
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ApplicationConsumerProjectRouteHash(project_hash=resource.project_hash,)


class ApplicationConsumerProjectRouteHashArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ApplicationConsumerProjectRouteHash.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ApplicationConsumerProjectRouteHash.from_proto(i) for i in resources]


class ApplicationDispatchRules(object):
    def __init__(self, domain: str = None, path: str = None, service: str = None):
        self.domain = domain
        self.path = path
        self.service = service

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = application_pb2.AppengineApplicationDispatchRules()
        if Primitive.to_proto(resource.domain):
            res.domain = Primitive.to_proto(resource.domain)
        if Primitive.to_proto(resource.path):
            res.path = Primitive.to_proto(resource.path)
        if Primitive.to_proto(resource.service):
            res.service = Primitive.to_proto(resource.service)
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ApplicationDispatchRules(
            domain=resource.domain, path=resource.path, service=resource.service,
        )


class ApplicationDispatchRulesArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ApplicationDispatchRules.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ApplicationDispatchRules.from_proto(i) for i in resources]


class ApplicationFeatureSettings(object):
    def __init__(
        self, split_health_checks: bool = None, use_container_optimized_os: bool = None
    ):
        self.split_health_checks = split_health_checks
        self.use_container_optimized_os = use_container_optimized_os

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = application_pb2.AppengineApplicationFeatureSettings()
        if Primitive.to_proto(resource.split_health_checks):
            res.split_health_checks = Primitive.to_proto(resource.split_health_checks)
        if Primitive.to_proto(resource.use_container_optimized_os):
            res.use_container_optimized_os = Primitive.to_proto(
                resource.use_container_optimized_os
            )
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ApplicationFeatureSettings(
            split_health_checks=resource.split_health_checks,
            use_container_optimized_os=resource.use_container_optimized_os,
        )


class ApplicationFeatureSettingsArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ApplicationFeatureSettings.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ApplicationFeatureSettings.from_proto(i) for i in resources]


class ApplicationIap(object):
    def __init__(
        self,
        enabled: bool = None,
        oauth2_client_id: str = None,
        oauth2_client_secret: str = None,
        oauth2_client_secret_sha256: str = None,
    ):
        self.enabled = enabled
        self.oauth2_client_id = oauth2_client_id
        self.oauth2_client_secret = oauth2_client_secret
        self.oauth2_client_secret_sha256 = oauth2_client_secret_sha256

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = application_pb2.AppengineApplicationIap()
        if Primitive.to_proto(resource.enabled):
            res.enabled = Primitive.to_proto(resource.enabled)
        if Primitive.to_proto(resource.oauth2_client_id):
            res.oauth2_client_id = Primitive.to_proto(resource.oauth2_client_id)
        if Primitive.to_proto(resource.oauth2_client_secret):
            res.oauth2_client_secret = Primitive.to_proto(resource.oauth2_client_secret)
        if Primitive.to_proto(resource.oauth2_client_secret_sha256):
            res.oauth2_client_secret_sha256 = Primitive.to_proto(
                resource.oauth2_client_secret_sha256
            )
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ApplicationIap(
            enabled=resource.enabled,
            oauth2_client_id=resource.oauth2_client_id,
            oauth2_client_secret=resource.oauth2_client_secret,
            oauth2_client_secret_sha256=resource.oauth2_client_secret_sha256,
        )


class ApplicationIapArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ApplicationIap.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ApplicationIap.from_proto(i) for i in resources]


class ApplicationParent(object):
    def __init__(self, type: str = None, id: str = None):
        self.type = type
        self.id = id

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = application_pb2.AppengineApplicationParent()
        if Primitive.to_proto(resource.type):
            res.type = Primitive.to_proto(resource.type)
        if Primitive.to_proto(resource.id):
            res.id = Primitive.to_proto(resource.id)
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ApplicationParent(type=resource.type, id=resource.id,)


class ApplicationParentArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ApplicationParent.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ApplicationParent.from_proto(i) for i in resources]


class ApplicationConsumerProjectServiceEnum(object):
    @classmethod
    def to_proto(self, resource):
        if not resource:
            return resource
        return application_pb2.AppengineApplicationConsumerProjectServiceEnum.Value(
            "AppengineApplicationConsumerProjectServiceEnum%s" % resource
        )

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return resource
        return application_pb2.AppengineApplicationConsumerProjectServiceEnum.Name(
            resource
        )[len("AppengineApplicationConsumerProjectServiceEnum") :]


class ApplicationDatabaseTypeEnum(object):
    @classmethod
    def to_proto(self, resource):
        if not resource:
            return resource
        return application_pb2.AppengineApplicationDatabaseTypeEnum.Value(
            "AppengineApplicationDatabaseTypeEnum%s" % resource
        )

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return resource
        return application_pb2.AppengineApplicationDatabaseTypeEnum.Name(resource)[
            len("AppengineApplicationDatabaseTypeEnum") :
        ]


class ApplicationServingStatusEnum(object):
    @classmethod
    def to_proto(self, resource):
        if not resource:
            return resource
        return application_pb2.AppengineApplicationServingStatusEnum.Value(
            "AppengineApplicationServingStatusEnum%s" % resource
        )

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return resource
        return application_pb2.AppengineApplicationServingStatusEnum.Name(resource)[
            len("AppengineApplicationServingStatusEnum") :
        ]


class Primitive(object):
    @classmethod
    def to_proto(self, s):
        if not s:
            return ""
        return s

    @classmethod
    def from_proto(self, s):
        return s
